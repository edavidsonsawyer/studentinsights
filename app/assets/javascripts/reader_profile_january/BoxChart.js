import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import chroma from 'chroma-js';
import {previousGrade, gradeText} from '../helpers/gradeText';
import {toSchoolYear} from '../helpers/schoolYear';
import {benchmarkPeriodToMoment} from '../reading/readingData';
import {shouldHighlight} from './dibelsParsing';
import {ORANGE, GREEN, PRESENT, BLANK} from './colors';


export default class BoxChart extends React.Component {
  render() {
    const {nowFn} = this.context;
    const {gradeNow, readerJson, benchmarkAssessmentKey} = this.props;
    const dataPoints = readerJson.benchmark_data_points.filter(d => d.benchmark_assessment_key === benchmarkAssessmentKey);
    const sortedDataPoints = _.sortBy(dataPoints, dataPoint => {
      return -1 * benchmarkPeriodToMoment(dataPoint.benchmark_period_key, dataPoint.benchmark_school_year).unix();
    });
    const schoolYear = toSchoolYear(nowFn());

    return (
      <div>
        <div style={styles.years}>
          <YearBox
            gradeThen={previousGrade(gradeNow)}
            schoolYear={schoolYear-1}
            sortedDataPoints={sortedDataPoints}
          />
          <YearBox
            gradeThen={gradeNow}
            schoolYear={schoolYear}
            sortedDataPoints={sortedDataPoints}
          />
        </div>
      </div>
    );
  }
}
BoxChart.propTypes = {
  readerJson: PropTypes.object.isRequired,
  benchmarkAssessmentKey: PropTypes.string.isRequired,
  gradeNow: PropTypes.string.isRequired
};
BoxChart.contextTypes = {
  nowFn: PropTypes.func.isRequired
};


// Take only most recent (the constraint comes from the input spreadsheets)
function YearBox(props) {
  const {schoolYear, gradeThen, sortedDataPoints} = props;
  const benchmarkPeriodKeys = ['fall', 'winter', 'spring'];
  const ds = benchmarkPeriodKeys.map(benchmarkPeriodKey => {
    return _.find(sortedDataPoints, d => d.benchmark_school_year === schoolYear && d.benchmark_period_key == benchmarkPeriodKey);
  });
  return (
    <div style={styles.yearBox}>
      <div style={styles.yearCells}>
        {benchmarkPeriodKeys.map((benchmarkPeriodKey, index) => {
          const dataPoint = ds[index];
          const valueEl = dataPoint ? dataPoint.json.value : null;
          const style = pickStyle(dataPoint, gradeThen);
          return (
            <div key={benchmarkPeriodKey} style={style} title={valueEl}>
              {benchmarkPeriodKey}
            </div>
          );
        })}
      </div>
      <div style={styles.yearWhen}>{gradeText(gradeThen)}, {schoolYear}</div>
    </div>
  );
}
YearBox.propTypes = {
  schoolYear: PropTypes.number.isRequired,
  gradeThen: PropTypes.string.isRequired,
  sortedDataPoints: PropTypes.array.isRequired
};


function pickStyle(dataPoint, gradeThen) {
  // no data
  if (!dataPoint) {
    return boxStyle(BLANK);
  }

  // data, but how should we color it?
  const isOrange = shouldHighlight(dataPoint, gradeThen);
  if (isOrange === true) {
    return boxStyle(ORANGE);
  }
  if (isOrange === false) {
    return boxStyle(GREEN);
  }

  // value, but no thresholds
  return boxStyle(PRESENT);
}


function boxStyle(color) {
  return {
    ...styles.box,
    background: color,
    outline: `1px solid ${chroma(color).darken().hex()}`,
    color: chroma(color).darken().darken().hex()
  };
}


const styles = {
  years: {
    display: 'flex',
    flexDirection: 'row',
    fontSize: 12
  },
  yearBox: {
    flex: 1,
    paddingRight: 20
  },
  yearCells: {
    display: 'flex',
    flexDirection: 'row'
  },
  yearWhen: {
    marginTop: 5
  },
  box: {
    flex: 1,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    height: 40,
    cursor: 'default'
  }
};
