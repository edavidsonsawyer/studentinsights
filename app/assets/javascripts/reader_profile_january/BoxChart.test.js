import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {withNowContext} from '../testing/NowContainer';
import PerDistrictContainer from '../components/PerDistrictContainer';
import {firstGradeWinter} from './BoxChart.fixture.js';
import BoxChart, {renderDibelsBoxFn, renderRawDibelsScoreBoxFn} from './BoxChart';


// this test data is only correct shape, but not semantically meaningful
export function testProps(props = {}) {
  return {
    gradeNow: '1',
    benchmarkAssessmentKey: 'f_and_p_english',
    readerJson: {
      access: {},
      services: [],
      iep_contents: null,
      feed_cards: [],
      "current_school_year": 2019,
      "benchmark_data_points": firstGradeWinter
    },
    renderBoxFn: renderDibelsBoxFn,
    ...props
  };
}

export function testRender(props = {}) {
  return withNowContext('2020-01-03T14:36:34.501Z',
    <PerDistrictContainer districtKey="somerville">
      <BoxChart {...props} />
    </PerDistrictContainer>
  );
}

it('renders without crashing', () => {
  const el = document.createElement('div');
  ReactDOM.render(testRender(testProps()), el);
});


it('snapshots renderDibelsBoxFn', () => {
  const tree = renderer.create(testRender(testProps())).toJSON();
  expect(tree).toMatchSnapshot();
});

it('snapshots renderRawDibelsScoreBoxFn', () => {
  const tree = renderer.create(testRender(testProps({renderBoxFn: renderRawDibelsScoreBoxFn}))).toJSON();
  expect(tree).toMatchSnapshot();
});