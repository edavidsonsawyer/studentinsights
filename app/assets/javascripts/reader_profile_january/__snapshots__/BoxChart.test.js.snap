// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 1`] = `
<div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "fontSize": 12,
      }
    }
  >
    <div
      style={
        Object {
          "flex": 1,
          "paddingRight": 20,
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        <div
          style={
            Object {
              "alignItems": "center",
              "background": "rgb(147, 196, 125)",
              "color": "#366525",
              "cursor": "default",
              "display": "flex",
              "flex": 1,
              "height": 40,
              "justifyContent": "center",
              "outline": "1px solid #649350",
            }
          }
          title="AA"
        >
          fall
        </div>
        <div
          style={
            Object {
              "alignItems": "center",
              "background": "orange",
              "color": "#904a00",
              "cursor": "default",
              "display": "flex",
              "flex": 1,
              "height": 40,
              "justifyContent": "center",
              "outline": "1px solid #c67600",
            }
          }
          title="B"
        >
          winter
        </div>
        <div
          style={
            Object {
              "alignItems": "center",
              "background": "rgb(147, 196, 125)",
              "color": "#366525",
              "cursor": "default",
              "display": "flex",
              "flex": 1,
              "height": 40,
              "justifyContent": "center",
              "outline": "1px solid #649350",
            }
          }
          title="E"
        >
          spring
        </div>
      </div>
      <div
        style={
          Object {
            "marginTop": 5,
          }
        }
      >
        Kindergarten
        , 
        2018
      </div>
    </div>
    <div
      style={
        Object {
          "flex": 1,
          "paddingRight": 20,
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
          }
        }
      >
        <div
          style={
            Object {
              "alignItems": "center",
              "background": "#ccc",
              "color": "#6d6d6d",
              "cursor": "default",
              "display": "flex",
              "flex": 1,
              "height": 40,
              "justifyContent": "center",
              "outline": "1px solid #9b9b9b",
            }
          }
          title="F"
        >
          fall
        </div>
        <div
          style={
            Object {
              "alignItems": "center",
              "background": "rgb(147, 196, 125)",
              "color": "#366525",
              "cursor": "default",
              "display": "flex",
              "flex": 1,
              "height": 40,
              "justifyContent": "center",
              "outline": "1px solid #649350",
            }
          }
          title="L"
        >
          winter
        </div>
        <div
          style={
            Object {
              "alignItems": "center",
              "background": "#f8f8f8",
              "color": "#959595",
              "cursor": "default",
              "display": "flex",
              "flex": 1,
              "height": 40,
              "justifyContent": "center",
              "outline": "1px solid #c5c5c5",
            }
          }
          title={null}
        >
          spring
        </div>
      </div>
      <div
        style={
          Object {
            "marginTop": 5,
          }
        }
      >
        1st grade
        , 
        2019
      </div>
    </div>
  </div>
</div>
`;
